Enable running tests with cargo-nextest

dburgener opened the pull request:
* Clean up CUSTOM_TEST_DIR appropriately after failures
* Set up nextest config to run api tests in serial

<!--
Thanks for your contribution!

We always run tests on the current stable version of Rust.
To avoid unexpected CI result, consider updating Rust if you're using an earlier version.

-->

Pat-Lafon commented:
As you set up nextest, did you notice a speedup or some other interesting benefit?

dburgener commented:
> As you set up nextest, did you notice a speedup or some other interesting benefit?

time cargo test:
real    0m2.500s
user    0m2.901s
sys     0m0.476s

time cargo nextest run:
real    0m1.154s
user    0m4.072s
sys     0m1.847s

So about twice as fast wall clock time (these are both after building everything, so should be just time to actually run the tests).

The nextest output also looks subjectively a little nicer to me?

I think the biggest benefit would just be all the configurability - things like running the API test in serial would have been a lot easier in nextest.  I left the mutex in for this PR, so that `cargo test` would continue to work, but the mutex is useless for nextest, since they are running in separate processes, so the mutex isn't actually shared.  The test group setup does the mutual exclusion there, and it's pretty trivial.  I don't have specific other use cases in mind, and this PR doesn't make much use of anything, but all the configuration and UI knobs seem like they could in theory be helpful.

Pat-Lafon commented:
I've tried cargo-nextest out and it's pretty slick. I'm a little sad I can't find a way to force cargo-nextest on `cargo test` invocation but that isn't too big of an issue. The output is a bit nicer which I appreciate. I saw something about nextest not supporting doc tests but I don't think we have any. In general, our ci testing times are probably more dependent on compilation than testing itself but the speed improvements are neat regardless.

The main reason to do this is still to enforce environment variable isolation.

dburgener commented:
I updated this to remove the interdependency with the `in_dir` code (and that fix is now part of #1082, where it belongs).  I also added a note in our CONTRIBUTING doc to clarify that both test runners are supported for now, but we are considering switching to exclusive `cargo-nextest` which is the plan I suggested [here](https://github.com/lalrpop/lalrpop/pull/1003#issuecomment-2931216331).

Pat-Lafon reviewed (approved):
LGTM

dburgener commented:
@Pat-Lafon just FYI, another issue I just noticed is that with regular cargo test, if one of the api tests fails, that means that it panics at that point and never calls the `teardown()` function.  That can cause other tests to failure because the state wasn't properly cleaned up (I was seeing it not change directory back to the original cwd).  Running these tests in `cargo nextest` seems to avoid that problem entirely.

I wouldn't be surprised if this is fixable in regular `cargo test`, but since we're talking about maybe mandating nextest anyways, we might not bother and consider it another point in favor of nextest.

Pat-Lafon commented:
That is a point in favor for sure. Would the suggestion here be to create wrap the Lock/State in a struct that calls teardown when it is dropped? I guess in the nextest case it doesn't matter as much because of the isolation and that each test creates a clean slate when it starts
